velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp

```diff
@@ -152,17 +152,25 @@ inline int32_t ArbitraryAggregate<int128_t>::accumulatorAlignmentSize() const {
   return static_cast<int32_t>(sizeof(int128_t));
 }
 
+struct ClusteredNonNumericAccumulator {
+  VectorPtr vector;
+  vector_size_t index;
+};
+
 // Arbitrary for non-numeric types. We always keep the first (non-NULL) element
 // seen. Arbitrary (x) will produce partial and final aggregations of type x.
 class NonNumericArbitrary : public exec::Aggregate {
  public:
   explicit NonNumericArbitrary(const TypePtr& resultType)
       : exec::Aggregate(resultType) {}
 
-  // We use singleValueAccumulator to save the results for each group. This
-  // struct will allow us to save variable-width value.
   int32_t accumulatorFixedWidthSize() const override {
-    return sizeof(SingleValueAccumulator);
+    return clusteredInput_ ? sizeof(ClusteredNonNumericAccumulator)
+                           : sizeof(SingleValueAccumulator);
+  }
+
+  bool accumulatorUsesExternalMemory() const override {
+    return true;
```

Comment:
Is this an unrelated fix?

Reply:
It's kind of related, the new `ClusteredNonNumericAccumulator` needs its destructor called, but `SingleValueAccumulator` actually does not, in most of the cases the memory is released along with the HSA.